### PR TITLE
esp32\machine_pwm.c: HOTFIX PWM duty.

### PR DESCRIPTION
**Pull request by IhorNehrutsa**
Save the same duty cycle when frequency (or frequency resolution) is changed.

The bug is detected with **Example of a smooth frequency change** from https://docs.micropython.org/en/latest/esp32/tutorial/pwm.html

### CODE REVIEW

File: ports/esp32/machine_pwm.c [hunk 1]
```diff
@@ -205,18 +219,15 @@ STATIC void configure_channel(machine_pwm_obj_t *self) {
...
-        PWM_DBG("set_freq(%d)", freq)
 
         // Find the highest bit resolution for the requested frequency
         unsigned int i = LEDC_APB_CLK_HZ; // 80 MHz
         if (freq < EMPIRIC_FREQ) {
             i = LEDC_REF_CLK_HZ; // 1 MHz
         }
 
-        #if 1
+        #if 0
```
Comment (dpgeorge) [inline ports/esp32/machine_pwm.c:230]: Can this small block of code be removed now that it's unused (the 2 lines below here)?

Comment (IhorNehrutsa) [reply ports/esp32/machine_pwm.c]: Will remove after using ESP-IDF5.

**Review by dpgeorge — DISMISSED** [review]

File: ports/esp32/machine_pwm.c [hunk 2]
```diff
@@ -77,6 +77,23 @@ STATIC ledc_timer_config_t timers[PWM_TIMER_MAX];
...
+#else
+#define HIGHEST_PWM_RES (LEDC_TIMER_14_BIT)
+#endif
+// duty resolution of user interface in `duty_u16()` and `duty_u16` parameter in constructor/initializer
+#define UI_RES_16_BIT (16)
+// how much to shift from the HIGHEST_PWM_RES duty resolution to the user interface duty resolution UI_RES_16_BIT
+#define UI_RES_SHIFT (16 - HIGHEST_PWM_RES) // 0 for ESP32, 2 for S2, S3, C3
+
+// If the PWM frequency is less than EMPIRIC_FREQ, then LEDC_REF_CLK_HZ(1 MHz) is used, else LEDC_APB_CLK_HZ(80 MHz) is used
+#define EMPIRIC_FREQ (10) // Hz
```
Comment (dpgeorge) [inline ports/esp32/machine_pwm.c:96]: It looks like this chunk of defines and comments is already present just below.

File: ports/esp32/machine_pwm.c [hunk 3]
```diff
@@ -316,23 +329,27 @@ STATIC void set_duty_u16(machine_pwm_obj_t *self, int duty) {
...
-    duty >>= HIGHEST_PWM_RES + UI_RES_SHIFT - timers[TIMER_IDX(self->mode, self->timer)].duty_resolution;
-    int max_duty = (1 << timers[TIMER_IDX(self->mode, self->timer)].duty_resolution) - 1;
-    if (duty < 0) {
-        duty = 0;
-    } else if (duty > max_duty) {
-        duty = max_duty;
-    }
-    check_esp_err(ledc_set_duty(self->mode, self->channel, duty));
+    ledc_timer_config_t timer = timers[TIMER_IDX(self->mode, self->timer)];
+    int _duty = duty >> (HIGHEST_PWM_RES + UI_RES_SHIFT - timer.duty_resolution);
```
Comment (dpgeorge) [inline ports/esp32/machine_pwm.c:333]: Identifiers starting with `_` are reserved for the system, please don't use them.

**Review by dpgeorge — CHANGES_REQUESTED** [review]

**Review by dpgeorge — APPROVED** [review]

**Comment by dpgeorge** [comment]
Merged in 09fe80d0912b2418fb704cf8e80db1712fb7523f

**Pull request closed without merging**